v3/src/components/map/models/map-grid-model.ts

```diff
@@ -0,0 +1,205 @@
+          if (long !== undefined && lat !== undefined) {
+            const longIndex = Math.floor((long - startWest) / gridSize),
+              latIndex = Math.floor((lat - startSouth) / gridSize)
+            rectArray.addCaseToRect(longIndex, latIndex, caseID)
+          }
+        })
+      }
+    }
+
+    return {
+      actions: {
+        initializeRectArray() {
+          const cases = self.dataConfiguration?.caseDataArray,
+            bounds = self.dataConfiguration ? getLatLongBounds(self.dataConfiguration) : undefined
+          if (cases && cases.length > 0 && bounds && bounds.isValid()) {
+            setupRectangles()
+            computeCounts()
+            self.rectArray.deleteZeroRects()
+          }
+        },
+        clear() {
+          self.rectArray.clear()
+        },
+        updateSelection() {
+          self.forEachRect((rect) => {
+            rect.selected = rect.cases.every((aCase) => self.dataConfiguration?.dataset?.isCaseSelected(aCase))
+          })
+        },
+      }
+    }
+  })
+  .actions(self => ({
+    afterCreate() {
+      addDisposer(self, reaction(
+        () => [self.dataConfiguration,
+          self.dataConfiguration?.hiddenCases.length, self.gridMultiplier, self._gridSize],
+        () => {
+          self.initializeRectArray()
+          self.updateSelection()
+        },
+        {name: "MapGridModel.afterCreate.reaction [dataConfiguration]", fireImmediately: true}
+      ))
+      addDisposer(self, reaction(
+        () => [self.dataConfiguration?.selection?.length],
+        () => self.updateSelection(),
+        {name: "MapGridModel.afterCreate.reaction [selection]"}
+      ))
+    },
+  }))
+  .views(self => ({
+    _selectCasesInRect(longIndex: number, latIndex: number, select: boolean, extend: boolean) {
+      const dataset = self.dataConfiguration?.dataset,
+        rect = self.rectArray.getRect(longIndex, latIndex)
+      if (rect) {
+        if (extend) {
+          dataset?.selectCases(rect.cases, select)
+        } else {
+          dataset?.setSelectedCases(rect.cases)
+        }
+      }
+    },
+    selectCasesInRect(longIndex: number, latIndex: number, select: boolean, extend: boolean) {
+      if (!extend) {
+        self.forEachRect((rect) => {
+          rect.selected = false
+        })
+      }
+      this._selectCasesInRect(longIndex, latIndex, select, extend)
+    },
+    deselectCasesInRect(longIndex: number, latIndex: number) {
+      this._selectCasesInRect(longIndex, latIndex, false, false)
+    },
+    deselectAll() {
+      self.dataConfiguration?.dataset?.selectAll(false)
+    },
+
+  }))
+  .actions(applyUndoableAction)
+
+export interface IMapGridModel extends Instance<typeof MapGridModel> {
+}
+
+export interface IMapGridModelSnapshot extends SnapshotIn<typeof MapGridModel> {
+}
+
+/*
+export function isMapGridModel(layerModel?: IMapLayerModel): layerModel is IMapGridModel {
+  return layerModel?.type === kMapGridType
+}
+*/
+
```

Comment:
Since the `MapGridModel` isn't polymorphic, I think you can remove the commented out code.